pkg/service/model_build_listener.go

```diff
@@ -0,0 +1,110 @@
+package service
+
+import (
+	"context"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	elbv2api "sigs.k8s.io/aws-load-balancer-controller/apis/elbv2/v1beta1"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/annotations"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/k8s"
+	elbv2model "sigs.k8s.io/aws-load-balancer-controller/pkg/model/elbv2"
+	"strconv"
+)
+
+func (t *defaultModelBuildTask) buildListeners(ctx context.Context, ec2Subnets []*ec2.Subnet) error {
+	tgAttrs, err := t.buildTargetGroupAttributes(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to build target group attributes")
+	}
+	modelTargetType := t.buildTargetType(ctx)
+	apiTargetType := elbv2api.TargetType(modelTargetType)
+	preserveClientIP, err := t.buildPreserveClientIPFlag(ctx, apiTargetType)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to get preserve Client IP configuration")
+	}
+	var certificateARNs []string
+	t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixSSLCertificate, &certificateARNs, t.service.Annotations)
+
+	var sslPorts []string
+	t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixSSLPorts, &sslPorts, t.service.Annotations)
+	sslPortsSet := sets.NewString(sslPorts...)
+
+	backendProtocol := ""
+	t.annotationParser.ParseStringAnnotation(annotations.SvcLBSuffixBEProtocol, &backendProtocol, t.service.Annotations)
+
+	targetGroupMap := map[string]*elbv2model.TargetGroup{}
+
+	for _, port := range t.service.Spec.Ports {
+		hc, err := t.buildTargetHealthCheck(ctx)
+		tgProtocol := elbv2model.Protocol(port.Protocol)
+		listenerProtocol := elbv2model.Protocol(port.Protocol)
+		if err != nil {
+			return err
+		}
+
+		if tgProtocol != elbv2model.ProtocolUDP && certificateARNs != nil && (sslPortsSet.Len() == 0 || sslPortsSet.Has(port.Name) || sslPortsSet.Has(strconv.Itoa(int(port.Port)))) {
+			if backendProtocol == "ssl" {
+				tgProtocol = elbv2model.ProtocolTLS
+			}
+			listenerProtocol = elbv2model.ProtocolTLS
+		}
+
+		svcPort := intstr.FromInt(int(port.Port))
+		tgName := t.buildTargetGroupName(ctx, port.TargetPort, modelTargetType, tgProtocol, hc)
+		tgResId := t.buildTargetGroupResourceID(k8s.NamespacedName(t.service), svcPort)
+		targetPort := 1
+		if port.TargetPort.Type == intstr.Int {
```

Comment:
1. let's have a method like `buildTargetGroupPort`. 
2. use the built targetGroupPort as hash component instead of `port.TargetPort` when construct tgName. since it's the actual immutable targetGroup attribute.

https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/ingress/model_build_target_group.go#L126